### PR TITLE
[FIX] mail: detect space inside channel name

### PR DESCRIPTION
**Pull request by phenix-factory**
task-2366680

### CODE REVIEW

File: addons/mail/static/src/models/composer/composer.js [hunk 1]
```diff
@@ -146,7 +146,7 @@ function factory(dependencies) {
                     recordReplacement = this[this.activeSuggestedRecordName].substitution;
                     break;
                 case 'activeSuggestedChannel':
-                    recordReplacement = this[this.activeSuggestedRecordName].name;
+                    recordReplacement = this[this.activeSuggestedRecordName].name.replace(/ /g, '\u00a0');
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:149]: missing test for the space in channel name case

File: addons/mail/static/src/models/composer/composer.js [hunk 2]
```diff
@@ -509,7 +509,7 @@ function factory(dependencies) {
          * @returns {mail.partner[]}
          */
         _computeMentionedChannels() {
-            const inputMentions = this.textInputContent.match(
+            const inputMentions = this.textInputContent.replace('\n', "\n ").match(
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:512]: missing test for channel name wrongly detected when followed directly by a new line

File: addons/mail/static/src/models/composer/composer.js [hunk 3]
```diff
@@ -563,7 +563,7 @@ function factory(dependencies) {
             let startIndex = 0;
             for (const match of inputMentions) {
                 const suggestionDelimiter = match[0];
-                const matchName = owl.utils.escape(match.substring(1).replace(new RegExp('\u00a0', 'g'), ' '));
+                const matchName = match.substring(1).replace(new RegExp('\u00a0', 'g'), ' ');
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:566]: Please rename the variable to `matchNameEscaped` if this is indeed already escaped from `body`. That will make the code less confusing.

File: addons/mail/static/src/models/composer/composer.js [hunk 4]
```diff
@@ -574,7 +574,7 @@ function factory(dependencies) {
                     cssClass = "o_channel_redirect";
                 }
                 const mention = this[field].find(mention =>
-                    mention.name === matchName
+                    owl.utils.escape(mention.name) === matchName
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:577]: Missing a test for correct channel detection with characters that can be escaped in the name.

**Review by seb-odoo — CHANGES_REQUESTED** [review]

**Comment by seb-odoo** [comment]
Also all commits should have the task-id (and ideally the PR number too then) + there are typo in 2nd commit message. (however, its, ...)

File: addons/mail/static/src/components/thread_view/thread_view_tests.js [hunk 5]
```diff
@@ -1181,6 +1181,168 @@ QUnit.test('Post a message containing an email address followed by a mention on
...
+    );
+});
+
+QUnit.test('mention a channel with new line in the name', async function (assert) {
+    assert.expect(2);
+
+    this.data['mail.channel'].records.push({
+        id: 7,
+        name: "General good",
+        public: "groups",
```
Comment (seb-odoo) [inline addons/mail/static/src/components/thread_view/thread_view_tests.js:1295]: same

File: addons/mail/static/src/components/thread_view/thread_view_tests.js [hunk 6]
```diff
@@ -1181,6 +1181,168 @@ QUnit.test('Post a message containing an email address followed by a mention on
...
+    });
+    this.data['res.partner'].records.push({
+        id: 25,
+        email: "testpartner@odoo.com",
+        name: "TestPartner",
+    });
```
Comment (seb-odoo) [inline addons/mail/static/src/components/thread_view/thread_view_tests.js:1301]: same

File: addons/mail/static/src/components/thread_view/thread_view_tests.js [hunk 7]
```diff
@@ -1181,6 +1181,168 @@ QUnit.test('Post a message containing an email address followed by a mention on
...
+        "message should contain a link the the mentionned channel"
+    );
+    assert.strictEqual(
+        document.querySelector('.o_channel_redirect').textContent,
+        '#General & good',
+        "link to the channel must containt # + the channel name"
+    );
+});
+
+QUnit.test('mention a channel with new line in the name', async function (assert) {
```
Comment (seb-odoo) [inline addons/mail/static/src/components/thread_view/thread_view_tests.js:1289]: This is not the goal of the test, the issue is when channel is followed by a new line in the body, not when its name contain a new line

File: addons/mail/static/src/components/thread_view/thread_view_tests.js [hunk 8]
```diff
@@ -1181,6 +1181,168 @@ QUnit.test('Post a message containing an email address followed by a mention on
...
+    });
+    await afterNextRender(() => {
+        document.querySelector(`.o_ComposerTextInput_textarea`)
+            .dispatchEvent(new window.KeyboardEvent('keydown'));
+        document.querySelector(`.o_ComposerTextInput_textarea`)
+            .dispatchEvent(new window.KeyboardEvent('keyup'));
+    });
+    await afterNextRender(() =>
+        document.querySelector('.o_ComposerSuggestion').click()
+    );
```
Comment (seb-odoo) [inline addons/mail/static/src/components/thread_view/thread_view_tests.js:1325]: same question

File: addons/mail/static/src/components/thread_view/thread_view_tests.js [hunk 9]
```diff
@@ -1181,6 +1181,168 @@ QUnit.test('Post a message containing an email address followed by a mention on
...
+    await afterNextRender(() => {
+        document.querySelector(`.o_ComposerTextInput_textarea`).focus();
```
Comment (seb-odoo) [inline addons/mail/static/src/components/thread_view/thread_view_tests.js:1327]: Isn't the focus put back automatically on the textarea on selecting a mention? Otherwise that would be an annoying bug

File: addons/mail/static/src/components/thread_view/thread_view_tests.js [hunk 10]
```diff
@@ -1181,6 +1181,168 @@ QUnit.test('Post a message containing an email address followed by a mention on
...
+        document.execCommand('insertText', false, "\n test");
+    });
+    console.log(document.querySelector(`.o_ComposerTextInput_textarea`).textContent);
```
Comment (seb-odoo) [inline addons/mail/static/src/components/thread_view/thread_view_tests.js:1330]: no log

**Review by seb-odoo — CHANGES_REQUESTED** [review]
See comments. Did you ensure the tests failed when reverting the corresponding fixes?

**Comment by phenix-factory** [comment]
@seb-odoo Yes the tests are failling without the fixs.

File: addons/mail/static/src/components/thread_view/thread_view_tests.js (hunk 9 above)
Comment (phenix-factory) [reply addons/mail/static/src/components/thread_view/thread_view_tests.js:1302]: You right

File: addons/mail/static/src/components/thread_view/thread_view_tests.js [hunk 11]
```diff
@@ -1181,6 +1181,141 @@ QUnit.test('Post a message containing an email address followed by a mention on
...
+        document.querySelector('.o_Message_content'),
+        '.o_channel_redirect',
+        "message should contain a link the the mentionned channel"
+    );
+    assert.strictEqual(
+        document.querySelector('.o_channel_redirect').textContent,
+        '#General good',
+        "link to the channel must containt # + the channel name"
+    );
+});
```
Comment (seb-odoo) [inline addons/mail/static/src/components/thread_view/thread_view_tests.js:1317]: I just realized I gave you the wrong setup for this test, and the one I intended to cover is actually still not working.

First here is the correct setup for this test to cover the fix you did. I'm pretty sure this test is currently succeeding even if you revert the `\n` fix.

It should be changed so that there is a first line that contains a `#` but it is not a mention.
And then the mention should be on the 2nd line (this would be similar to the existing partner test).



And now the tests I had in mind, that are still not working (neither for partner or channel):

Have a mention directly followed by a `\n`, by **replacing** the space that is automatically added after selecting a mention by a `\n`. This bug is due to using space as separator being really weak.

(And a similar version but instead of a \n it's the end of the input).

Basically `_computeMentionedPartners` and `_computeMentionedChannels`, instead of looking for things that look like mentions with regex and then matching them to channel/partner name, should look for known partner/channel name (+ the mention starting char) inside the text input.

Comment (phenix-factory) [reply addons/mail/static/src/components/thread_view/thread_view_tests.js:1317]: > Have a mention directly followed by a \n, by replacing the space that is automatically added after selecting a mention by a \n. This bug is due to using space as separator being really weak.

This case is not fixed by the current code. I'll build the test and try to find a solution.

**Comment by seb-odoo** [comment]
You will need to rebase on https://github.com/odoo/odoo/pull/61119 too

**Comment by seb-odoo** [comment]
I pushed a commit where I removed the regex as I suggested earlier. Let's see if runbot likes it.

**Comment by seb-odoo** [comment]
Also done changes to ensure we can mention different partners/channel that have the same name with new code + added tests to cover those.

**Comment by seb-odoo** [comment]
I'll let @alexkuhn review this at his return since I have done most changes now.

**Review by alexkuhn — APPROVED** [review]
LGTM

**Comment by alexkuhn** [comment]
@robodoo r+ rebase-ff
